tfhe/src/shortint/server_key/tests/noise_level.rs

```diff
@@ -0,0 +1,263 @@
+use crate::shortint::ciphertext::NoiseLevel;
+use crate::shortint::keycache::KEY_CACHE;
+use crate::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+use crate::shortint::{Ciphertext, ServerKey};
+
+fn test_1_ct_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext) {
+    let test_fn = |f: &dyn Fn(&ServerKey, &Ciphertext) -> Ciphertext,
+                   g: &dyn Fn(NoiseLevel) -> NoiseLevel| {
+        assert!(f(sk, ct).noise_level == g(ct.noise_level));
+    };
+
+    test_fn(&ServerKey::unchecked_neg, &|a| a);
+    test_fn(
+        &|sk, ct| ServerKey::unchecked_neg_with_correcting_term(sk, ct).0,
+        &|a| a,
+    );
+
+    let acc = sk.generate_lookup_table(|_| 0);
+
+    test_fn(
+        &|sk, ct| ServerKey::apply_lookup_table(sk, ct, &acc),
+        &|_| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_1_ct_assign_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext) {
+    let test_fn = |f: &dyn Fn(&ServerKey, &mut Ciphertext),
+                   g: &dyn Fn(NoiseLevel) -> NoiseLevel| {
+        let mut clone = ct.clone();
+        f(sk, &mut clone);
+        assert!(clone.noise_level == g(ct.noise_level));
+    };
+
+    test_fn(&ServerKey::unchecked_neg_assign, &|a| a);
+    test_fn(
+        &|sk, ct| {
+            ServerKey::unchecked_neg_assign_with_correcting_term(sk, ct);
+        },
+        &|a| a,
+    );
+
+    let acc = sk.generate_lookup_table(|_| 0);
+
+    test_fn(
+        &|sk, ct| ServerKey::apply_lookup_table_assign(sk, ct, &acc),
+        &|_| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_2_ct_noise_level_propagation(sk: &ServerKey, ct1: &Ciphertext, ct2: &Ciphertext) {
+    let test_fn = |f: &dyn Fn(&ServerKey, &Ciphertext, &Ciphertext) -> Ciphertext,
+                   g: &dyn Fn(NoiseLevel, NoiseLevel) -> NoiseLevel| {
+        assert!(f(sk, ct1, ct2).noise_level == g(ct1.noise_level, ct2.noise_level));
+    };
+
+    test_fn(&ServerKey::unchecked_add, &|a, b| a + b);
+    test_fn(&ServerKey::unchecked_sub, &|a, b| a + b);
+    test_fn(
+        &|sk, ct1, ct2| ServerKey::unchecked_sub_with_correcting_term(sk, ct1, ct2).0,
+        &|a, b| a + b,
+    );
+
+    let any_trivially_encrypted = ct1.degree.0 == 0 || ct2.degree.0 == 0;
```

Comment:
isn't this better detected with the Noise level being 0 ?

Reply:
IMO it's better to use the same condition as in the op themselves
Otherwise the assert may not pass

Reply:
ok but it means we may need to progressively improve the operations handling of trivial having a fn is_trivial on ciphertexts checking their noise, can you open a follow up issue please ?

Reply:
with a comment on those checks with a TODO

Reply:
The ops multiplication optimization needs one of the encrypted value to be `0` which is independent (a priori) from its noise level to be 0.
Other optimizations of trivial encryptions could be interesting but not in this case I think

Reply:
Did we open a follow up issue to have a is_trivial function on shortint ?

Reply:
if we don't have an issue we need one, and yes to keep the check consistent with ops for now